datastore/complete_post_object.go

```diff
@@ -0,0 +1,86 @@
+package datastore
+
+import (
+	"fmt"
+
+	nex "github.com/PretendoNetwork/nex-go"
+	common_globals "github.com/PretendoNetwork/nex-protocols-common-go/globals"
+	datastore "github.com/PretendoNetwork/nex-protocols-go/datastore"
+	datastore_types "github.com/PretendoNetwork/nex-protocols-go/datastore/types"
+)
+
+func completePostObject(err error, client *nex.Client, callID uint32, param *datastore_types.DataStoreCompletePostParam) uint32 {
+	if commonDataStoreProtocol.s3ObjectSizeHandler == nil {
+		common_globals.Logger.Warning("S3ObjectSize not defined")
+		return nex.Errors.Core.NotImplemented
+	}
+
+	if commonDataStoreProtocol.getObjectSizeDataIDHandler == nil {
+		common_globals.Logger.Warning("GetObjectSizeDataID not defined")
+		return nex.Errors.Core.NotImplemented
+	}
+
+	if commonDataStoreProtocol.updateObjectUploadCompletedByDataIDHandler == nil {
+		common_globals.Logger.Warning("UpdateObjectUploadCompletedByDataID not defined")
+		return nex.Errors.Core.NotImplemented
+	}
+
+	if err != nil {
+		common_globals.Logger.Error(err.Error())
+		return nex.Errors.DataStore.Unknown
+	}
+
+	// TODO - What is param.IsSuccess? Is this correct?
+	if param.IsSuccess {
+		bucket := commonDataStoreProtocol.s3Bucket
+		key := fmt.Sprintf("%d.bin", param.DataID)
+
+		objectSizeS3, err := commonDataStoreProtocol.s3ObjectSizeHandler(bucket, key)
+		if err != nil {
+			common_globals.Logger.Error(err.Error())
+			return nex.Errors.DataStore.NotFound
+		}
+
+		objectSizeDB, errCode := commonDataStoreProtocol.getObjectSizeDataIDHandler(param.DataID)
```

Comment:
It provides a much quicker way to get the objects size, which is needed when sending download URLs to the client, since it can be retrieved with a single SQL query rather than querying S3 every time. Also it allows us to verify that the object posted correctly/completely, just in case some weird edge case happens where the upload only partially completes

Reply:
I see. In that case, I think we should log an error when the S3 size and the local size doesn't match